issue-3381: preventing validation occurring prior to 'beforeChange'

jdmcnair opened the pull request:
### Context
Issue described in https://github.com/handsontable/handsontable/issues/3381. Validation happens on changes before the potential cancelling of those changes in the `beforeChange` event. This means that the table may sometimes show validation unrelated to the actual data in the table.

### How has this been tested?
New tests in `Core_validate.spec.js`

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Related issue(s):
1. 3381
2.
3.

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the code style of this project,
- [ ] My change requires a change to the documentation.

AMBudnik commented:
Thank you for sharing this fix. Please change the branch to `develop` and the PR will be ready for a dev review.

jdmcnair commented:
@AMBudnik I've changed the target branch to `develop`. Note that this request differential now contains a few minor commits from the master branch that I had used as a base. Mostly merges, and one documentation change. They seem pretty harmless, but let me know if those need to be cleared out. I'm sure I can work out how to do so if it's worth the effort.

AMBudnik commented:
Thank you so much @jdmcnair 

I am really grateful. You have been so open for sharing your work investigation.

TheLudd commented:
Will this change also make sure that if changes are added/removed through `beforeChange`, then those are reflected in the validation? 

I.e. I have a cell marked as invalid. When the user makes a change in another cell, I can sometimes update the invalid value through that change and I do this by modification to the changes in `beforeChange`. Today, the value gets updated but the cell is still invalid. Will this change fix that?

jdmcnair commented:
@TheLudd I'm not positive, but I don't think this will fix that. That seems like a separate issue where changes that are made programmatically during the `beforeChange` event do not trigger a re-validation of the altered cells.

TheLudd commented:
I just have the expectation that when changes are made, and you can modify those changes then validation should be done *after* the modification. This approach would also fit with your need which is that if changes are cancelled, no validation is needed since there are no changes.

AMBudnik commented:
Hi @TheLudd 
the current PR has a file conflict. Can you check what is causing it?

TheLudd commented:
It is not my PR

Den 29 jan. 2018 14:43 skrev "Aleksandra Budnik" <notifications@github.com>:

> Hi @TheLudd <https://github.com/theludd>
> the current PR has a file conflict. Can you check what is causing it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/handsontable/handsontable/pull/4509#issuecomment-361248964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAiA782GWe2Op3wX3aBGmuz8jzWxuwWjks5tPcsZgaJpZM4PRzIB>
> .
>

AMBudnik commented:
@TheLudd sorry, my bad.

jdmcnair commented:
@AMBudnik the conflict looks easy enough to resolve, superficially. Looks like there's an added line from another commit somewhere in the vicinity of my changes. I plan to take a closer look and make sure all the tests are still passing this week.

AMBudnik commented:
Thank you @jdmcnair 
And sorry for keeping you waiting so long for the code review.

jdmcnair commented:
@AMBudnik I've resolved the merge conflict and I've run through the unit tests. I can say for certain that the unit tests I added for this change are still passing. There is one failure in `contextMenu.e2e.js` that seems pretty far removed from the functionality I changed in this PR. The failing test is:

> should work properly (remove row) after destroy and new init

I tend to think that failure is unrelated to the changes in my PR or any residual merge conflicts. Let me know if there's anything else I need to do from here, but I think this PR should be ready to go again.

AMBudnik commented:
Thank you for the investigation @jdmcnair 
I am not capable to review the code but I have already changed the label. Hopefully, our dev team will be able to check this PR soon. We are planning to close all pending PR until the end of February.

AMBudnik commented:
I am so sorry. It looks awful that I wrote that we will check the PR on February, not we have May and there is no update. I am happy, however, that there aren't any file conflicts.

I have asked our CTO to schedule this PR check. If we're not able to plan them all to be checked let's do it with the smaller steps.

swistach commented on src/core.js:
```diff
@@ -1041,6 +1041,8 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
   }
 
   function validateChanges(changes, source, callback) {
+    verifyChanges();
```
Do you think that separated synchronous function is necessary? Maybe we can achieve the same effect using a body of `verifyChanges()` function directly in `validateChanges`. What's your opinion of my idea?

swistach commented on test/e2e/Core_validate.spec.js:
```diff
@@ -727,6 +727,84 @@ describe('Core_validate', () => {
     }, 200);
   });
 
+  it('should not add class name `htInvalid` for cancelled changes - on edit', (done) => {
```
Could you use `async` + `await sleep(...)` instead of `setTimeout(...)`?

swistach commented on test/e2e/Core_validate.spec.js:
```diff
@@ -727,6 +727,84 @@ describe('Core_validate', () => {
+    }, 500);
+  });
+
+  it('should not remove class name `htInvalid` for cancelled changes - on edit', (done) => {
```
Could you use `async` and `await sleep(...)` instead of `setTimeout(...)`?

AMBudnik commented:
Hi @jdmcnair 

is there any change for you to help us with this PR. Without adding the proposed changed (made by @swistach) we won't be able to pull the PR

jdmcnair commented:
@AMBudnik @swistach apologies on taking so long to respond to the above request for modifications. I was trying to sort out access to this repository, because I no longer belong to the surgeforward organization that created the fork that this PR was generated on. I wasn't able to secure access to make those changes, so I've opened up a new PR, #5557, for this same set of changes. I've included both the original changes proposed here, and the modifications requested above.
